### PR TITLE
Fix progress bar text orientation and scale

### LINKED ISSUE #14953: Text inside skinnable components should keep orientation rather than rotate or flip along with the component
This should probably be implemented in a reusable way, maybe as an extension method which setups up a scheduler on the drawable to keep it upright.

### Discussed in https://github.com/ppy/osu/discussions/14947

<div type='discussions-op-text'>

<sup>Originally posted by **CheeseManFuu** October  4, 2021</sup>
Super minor thing that I couldn't help but notice when playing with the UI Editor was that the progress bar's times and percentage were affected by transformations whereas the timing bar had its truck and bike icons scale and be unaffected by rotation and flipping. To have the progress bar on the top (and still move left-to-right) means flipping it vertically, which consequentially has the numbers also flipped vertically, and difficult to read. Simiarly, if you want the progress bar to be shorter or taller but the same width, to keep it out of the way of gameplay as possible but still visible, or less wide or wider but the same height, such as for people who use 4:3 or 21:9 resolutions, the text is squished or stretched. 

When it comes to stretching the element taller or wider, the text should remain the size it would be at the correct proportions, just be anchored to their respective corners.

A couple example screenshots and accompanying mock-ups of the fix below:

osu_2021-10-03_18-01-48
Normal on the left
osu_2021-10-03_18-01-48edit
Fixed on the left
osu_2021-10-03_18-02-49
Normal on the top
osu_2021-10-03_18-02-49edit
Fixed on the top</div>

### PR DESCRIPTION
**Pull request by HiddenNode**
Closes #14953.
fixes the text inside progress bar upright and prevents it from being rotated, scaled or flipped with the progress bar.

### CODE REVIEW

**Comment by bdach** [comment]
This solution has the exact same issue as the [previous attempt to fix this issue](https://github.com/ppy/osu/pull/18076):

https://user-images.githubusercontent.com/20418176/182474826-36f10da3-19ba-4733-9972-861f001be640.mp4

The text should not be sheared like that after the operations performed on the video.

**Comment by HiddenNode** [comment]
@bdach I made another implementation which seems to be working properly.

File: osu.Game/Extensions/DrawableExtensions.cs [hunk 1]
```diff
@@ -79,5 +80,42 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
...
+            parentMatrix.M13 = 0.0f;
+            parentMatrix.M23 = 0.0f;
+
+            if ((Math.Abs(Math.Abs(angle) - 90.0)) < 2.0f)
+            {
+                Matrix3 m = Matrix3.CreateRotationZ(MathHelper.DegreesToRadians(40.0f));
+                m.Transpose();
+                parentMatrix *= m;
+                drawable.Rotation = 40.0f;
+            }
```
Comment (bdach) [inline osu.Game/Extensions/DrawableExtensions.cs:104]: What's happening here? What's the check for, why is the threshold 2? Why 40 degrees?

File: osu.Game/Extensions/DrawableExtensions.cs [hunk 2]
```diff
@@ -79,5 +80,42 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
...
+                drawable.Rotation = 0.0f;
+
+            Matrix3 C = parentMatrix.Inverted();
+
+            float alpha, beta, sx, sy;
+            sy = C.M22;
+            alpha = C.M12 / C.M22;
+
+            beta = (C.M21 == 0.0f) ? 0.0f : 1 / ((C.M11 / C.M21) - alpha);
+            sx = (beta == 0.0f) ? C.M11 : C.M21 / beta;
```
Comment (bdach) [inline osu.Game/Extensions/DrawableExtensions.cs:115]: This needs to be documented better. I assume you're doing something like decomposing affine transforms here but it's unclear what is going on exactly.

File: osu.Game/Graphics/Containers/GrowToFitContainer.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Graphics.Containers;
+
+namespace osu.Game.Graphics.Containers
+{
+    /// <summary>
+    /// A container that grows in size to fit its child and retains its size when its child shrinks
```
Comment (bdach) [inline osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: Why is this needed?

File: osu.Game/Screens/Play/HUD/SongProgressInfo.cs [hunk 4]
```diff
@@ -106,5 +143,12 @@ protected override void Update()
         }
 
         private string formatTime(TimeSpan timeSpan) => $"{(timeSpan < TimeSpan.Zero ? "-" : "")}{Math.Floor(timeSpan.Duration().TotalMinutes)}:{timeSpan.Duration().Seconds:D2}";
+
+        private void keepTextSpritesUpright()
+        {
+            timeCurrentContainer.OnUpdate += Extensions.DrawableExtensions.KeepUprightAndUnscaled;
+            progressContainer.OnUpdate += Extensions.DrawableExtensions.KeepUprightAndUnscaled;
+            timeLeftContainer.OnUpdate += Extensions.DrawableExtensions.KeepUprightAndUnscaled;
```
Comment (bdach) [inline osu.Game/Screens/Play/HUD/SongProgressInfo.cs:151]: I'm not a huge fan of how this method needs to be attached to drawables via such hooks. I much preferred what #18076 did - namely, it introduced a special container type that would keep itself and its children upright. I would much prefer to use something like that. You could probably even co-opt `GrowToFitContainer` for this (but probably after a rename).

**Review by bdach — COMMENTED** [review]
The implementation does look to be working better, but I have several questions as to precisely _why_.

File: osu.Game/Extensions/DrawableExtensions.cs (hunk 2 above)
Comment (HiddenNode) [reply osu.Game/Extensions/DrawableExtensions.cs:115]: in the previous implementation, the Rotation and Scale of the drawable that we want to keep upright and unscaled is set to the inverse of the rotation and scale of its parent, which doesn't work because R(theta)*S(s)*R(-theta)*S(1/s) is not equal to the identity matrix, this implementation uses shear and scale instead of rotation and scale.

in the lines you mentioned sx and sy are the scale factors, alpha and beta are shear factors and the values being asingned are the solution of the system of equations that we need to solve to find the right factors.

File: osu.Game/Extensions/DrawableExtensions.cs (hunk 1 above)
Comment (HiddenNode) [reply osu.Game/Extensions/DrawableExtensions.cs:104]: As i mentioned in my reply below this implementation applies shear and scale to the drawable to revert the rotation and scale applied to the parent container, but it we can only revert the rotation and scale applied to the parent container with a shear and scale if the rotation is not equal to (90 or -90), in order to work around that we rotate the drawable a certain amount (it does not matter how much we rotate it all that matters is that we do not rotate it a multiple of 180 and i picked the number 40 randomly), and we adjust parentMatrix we that rotation.

In line 98 we are not only checking if the angle applied to the parent is not 90 or -90, we are also avoiding another problem, that is the more the rotation approaches 90 or -90 the bigger or smaller the scale and shear factors get, which causes a bug (the drawable disappears, and its update function does not get called again), i think that the drawable disappears because the factors being too big and too small is causing the floating point operations to be inaccurate.

I introduced a threshold to avoid that bug, i could have picked 1 degrees (or any small number) as the threshold but i picked 2 degrees just to be safe.

File: osu.Game/Graphics/Containers/GrowToFitContainer.cs (hunk 3 above)
Comment (HiddenNode) [reply osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: Previously (before the commit in which i put each TextSprite inside a GrowToFitContainer) the structure of SongProgressInfo was something like this:
SongProgressInfo:
    Container
        TextSprite
    Container
        TextSprite
    Container
        TextSprite
And each container has AutoSizeAxes = Axes.Both, so its size depended on the size of the TextSprite inside it, on the other hand each TextSprite is anchored to the center of the container and its origin is set to center, this means that the position of TextSprite depends on the size of the Container and the size of the Container depends on the size of the TextSprite, all of this would have been fine if the size of the TextSprite was fixed but it is not, as it depends on the text inside it which changes, causing indirectly the position of the TextSprite to change, in order to fix this it is necessary to fix the size of TextSprite which is a bad idea, so instead of doing that i put the TextSprites inside a container (GrowToFitContainer) that does not change its size every second and made it grow to fit any TextSprite size

File: osu.Game/Screens/Play/HUD/SongProgressInfo.cs (hunk 4 above)
Comment (HiddenNode) [reply osu.Game/Screens/Play/HUD/SongProgressInfo.cs:151]: Okay i will make the necessary changes, but i explained above GrowToFitContainer has its own purpose, which means that for KeepUprightContainer to be reusable it can't be mixed with GrowToFitContainer.

Comment (HiddenNode) [reply osu.Game/Screens/Play/HUD/SongProgressInfo.cs:151]: Also what do you think about calling KeepUprightAndUnscaled from inside UprightContainer's update instead of making it a member of UprightContainer so that it can be used in either way?

**Comment by HiddenNode** [comment]
@bdach I have changed and simplified `KeepUprightAndUnscaled`, added comments and removed the confusing if statement. I have also changed `SongProgressInfo` to use `UprightUnscaledContainer`.

File: osu.Game/Graphics/Containers/GrowToFitContainer.cs (hunk 3 above)
Comment (HiddenNode) [reply osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: You can see the problem if you remove the `GrowToFitContainer` and put the the `TextSprite`s directly inside the `Container` then go to skin editor and rotate the progress bar 45 degrees, i will upload a video showing this problem.

Comment (HiddenNode) [reply osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: https://user-images.githubusercontent.com/101936124/183753341-fe309639-fc13-43f7-ab3f-a30d74bdd361.mp4

Comment (HiddenNode) [reply osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: Here is a closer look: `TextSprite` is yellow, `Container` is red, `SongProgressBarInfo` is violet

https://user-images.githubusercontent.com/101936124/183762862-9e63a55f-b9c8-4914-8833-0f1a5e8ea7c4.mp4

t

Comment (HiddenNode) [reply osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: Notice how the size of `TextSprite` depends on the text inside it, which is affecting its position.

Comment (bdach) [reply osu.Game/Graphics/Containers/GrowToFitContainer.cs:10]: Well, on a quick look I unfortunately can't see a way to get rid of this `GrowToFitContainer`. But it shouldn't be required to manually instantiate one to have `UprightUnscaledContainer` work correctly. `UprightUnscaledContainer` itself should ensure that actual content is always nested in a `GrowToFitContainer`, and `GrowToFitContainer` should be a private inner class inaccessible by anything else.

There are [examples in the repository](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Catch/UI/CatchPlayfieldAdjustmentContainer.cs) of how you'd do this - you'd generally have the `GrowToFitContainer` as the `InternalChild` of the `UprightUnscaledContainer`, and override `Content` so that it points at the `GrowToFitContainer`, ensuring that any drawable that is `.Add()`ed to the `UprightUnscaledContainer` is actually internally added to the `GrowToFitContainer`.

File: osu.Game/Extensions/DrawableExtensions.cs [hunk 5]
```diff
@@ -79,5 +80,40 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
                     container.Add(child.CreateInstance());
             }
         }
+
+        /// <summary>
+        /// Keeps the drawable upright and unstretched preventing it from being rotated, sheared, scaled or flipped with its Parent.
+        /// </summary>
+        /// <param name="drawable">The drawable.</param>
+        public static void KeepUprightAndUnscaled(this Drawable drawable)
```
Comment (bdach) [inline osu.Game/Extensions/DrawableExtensions.cs:88]: Given that `KeepUprightAndUnscaled()` is basically useless without a `GrowToFitContainer`, I see no point in making it a publically-accessible method. It should be inside `UprightAndUnscaledContainer` as private.

File: osu.Game/Extensions/DrawableExtensions.cs [hunk 6]
```diff
@@ -79,5 +80,40 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
...
+        {
+            // Decomposes the inverse of the parent FrawInfo.Matrix into rotation, shear and scale.
+            var parentMatrix = drawable.Parent.DrawInfo.Matrix;
+            parentMatrix.Transpose();
```
Comment (bdach) [inline osu.Game/Extensions/DrawableExtensions.cs:92]: What does the transposition do here?

**Review by bdach — COMMENTED** [review]
There are failing code quality inspections in these latest commits.

**Comment by bdach** [comment]
About the requests to change API here: my view is essentially that making a drawable upright and unscaled should be *just* as easy as instantiating an `UprightAndUnscaledContainer`. It shouldn't require nesting `GrowToFitContainer`s, it shouldn't require installing manual `OnUpdate` hooks. Given the amount of setup involved I think it is essential to hide the underlying complexity from consumers of the container. To any next developer, this container should "just work".

**Comment by HiddenNode** [comment]
I have made the requested changes: `keepUprightAndUnscaled` and `GrowToFitContainer` are now both private members of `UprightAndUnscaledContainer`, I have also modified `GrowToFitContainer` to support multiple children, and removed the unnecessary transposition in `keepUprightAndUnscaled`.

File: osu.Game/Extensions/DrawableExtensions.cs (hunk 6 above)
Comment (HiddenNode) [reply osu.Game/Extensions/DrawableExtensions.cs]: this transposition is not necessary (i only used it to make `parrentMatrix` a row-major matrix, because i find row-major matrices more intuitive to work with), i am removing it.

**Comment by peppy** [comment]
What's the rationale for having the content not scale? That seems counter-intuitive from a user's perspective. Am I missing something?

**Comment by HiddenNode** [comment]
there is no rationale, i will change `UprightAndUnscaledContainer` to scale while preserving aspect ratio.

**Comment by HiddenNode** [comment]
I am basing the scale of text on the vertical scale of the progress bar
This is what it looks like:

https://user-images.githubusercontent.com/101936124/186259252-3dab19e6-2826-4070-9aeb-0f7aced8bc39.mp4

This is what it looks like with 50% scaling (i.e. if the progress bar is scaled vertically to 150% then the text would be scaled to 100 + (150 - 100) / 2 = 125% and if the progress bar is scaled vertically to 50% then the text would be scaled to 75%) :

https://user-images.githubusercontent.com/101936124/186258505-57b1ff6a-50ee-4344-84fb-c0899be9fcbd.mp4

With 50% scaling the text does not get too big or too small.
which one should i use? should i add a max/min scale (to prevent the text from getting too big/small)?

**Comment by peppy** [comment]
I'll repeat: why is the scale changing at all? Is there a requirement for this coming from somewhere I'm unaware of?

**Comment by HiddenNode** [comment]
The scale of the text has to change to prevent it from being stretched or squished as disscussed in #14947.

**Comment by HiddenNode** [comment]
This is what resizing/rescaling the progress bar in the skin editor currently (on the master branch) looks like:

https://user-images.githubusercontent.com/101936124/186410157-17da56cb-a55c-4936-ba65-ecad03c89138.mp4

**Comment by peppy** [comment]
I see.

Basing on height seems good for now, as you have it.

**Comment by HiddenNode** [comment]
I have changed the name of `UprightUnscaledContainer` to `UprightUnstretchedContainer`, and introduced two Properties to control how `UprightUnstretchedContainer` is scaled as an effort to make it reusable:
- `Scaling` controls whether `UprightUnscaledContainer` gets scaled or not, and which parent axis does `UprightUnscaledContainer` scale with (the default is to not scale).
- `ScalingFactor` controls how much `UprightUnscaledContainer` is scaled compared to its parent ( default is 1.0 which means that it scale the same amount).

I have set `ScalingFactor` to 0.5 for the text inside `ProgressBarInfo` to prevent it from obscuring the gameplay too much and to allow more space for the progress graph.

**Comment by peppy** [comment]
Please make a test scene showing this with all its combinations. Right now it's very hard to test (and several of the `enum` modes you added aren't used anywhere and thus not tested).

I'd imagine a grid test scene (see usages of `OsuGridTestScene`) which shows containers being rotated and scaled, with an `UprightAspectMaintainingContainer` in each, showing how it reacts to the different parents.

File: osu.Game/Graphics/Containers/UprightUnscaledContainer.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                    break;
+            }
+
+            usedScale = 1.0f + (usedScale - 1.0f) * ScalingFactor;
+
+            Scale = new Vector2(sx * usedScale, sy * usedScale);
+        }
+
+        /// <summary>
+        /// A container that grows in size to fit its children and retains its size when its children shrink
```
[the remark this answers is not in the document]
Comment (HiddenNode) [reply osu.Game/Graphics/Containers/UprightUnscaledContainer.cs]: I have tried explaining why this is required [here](https://github.com/ppy/osu/pull/19556#discussion_r939570458), that being said i will try to find a way to get rid of it.

**Comment by HiddenNode** [comment]
I have added 2 test scenes:
- `TestSceneUprightAspectMaintainingContainer`: tests `UprightAspectMaintainingContainer` with all combinations of rotation, scale and flip, and all combinations of `Scaling` and `ScaleFactor` parameters.
- `TestSceneGrowToFitContent`: tests (and showcases) the parameter `GrowToFitContent` which is used to toggle the behaviour of `GrowToFitContainer`.

I have also fixed some issues i found while making the test scenes.

**Comment by HiddenNode** [comment]
`UprightAspectMaintainingContainer` is not ready for review as i am still making changes to it.

**Comment by HiddenNode** [comment]
After much consideration, i have decided to remove `GrowToFitContainer` from `UprightAspectMaintainingContainer` because I realised while making the visual test that it was causing `UprightAspectMaintainingContainer` to function improperly in certain scenarios.
`GrowToFitContainer` was used to fix the issue shown [here](https://github.com/ppy/osu/pull/19556#discussion_r941817220), in order to avoid this issue, children of `UprightAspectMaintainingContainer` must not frequently change in size (as is the case for the `SpriteText`s used inside the SongProgressInfo which change in size every second to show a different time), so one way to fix this problem would be to set the size of the `SpriteText`s to a fixed size explicitly, but chosing the right size for the `timeLeft` and `timeCurrent` `SpriteText`s is very tricky because it depends on the beatmap length and font used in the `SpriteText`.
To solve this problem i have implemented `SizePreservingSpriteText` which maintains its maximum size effectively reducing how many times it changes in size, I have also added a test scene to show a comparison between `SizePreservingSpriteText` and `SpriteText` both of which are inside an `UprightAspectMaintainingContainer` to "simulate" the issue that I am talking about and show how `SizePreservingSpriteText` is used to circumvent it.

**Comment by peppy** [comment]
Thanks for the test coverage. I reduced the number of steps slightly as it was taking 15s+ to run. It should still give enough coverage to be assured things are working as expected.

**Review by peppy — APPROVED** [review]
